### DISCUSSION TITLE
foreach loop could behave like for loops for IList values

### DISCUSSION THREAD

**Discussion by VGib** [dotnet/csharplang#2586] [Language Ideas]
Hello,

This is the continuation of this [issue](https://github.com/dotnet/corefx/issues/38302)

**What is my problem**
I get big performance increase when dealing with calculating code with lot of arrays, which such a method (I just check that the IEnumerable is IList, since this has been lot done in Linq code)

```csharp
 public static void ForEach<T>(this IEnumerable<T> values, Action<T> doOnValue)
        {
            var valuesAsList = values as IList<T>;
            if (valuesAsList is null)
            {
                foreach (var value in values)
                {
                    doOnValue?.Invoke(value);
                }
            }
            else
            {
                for (int position = 0; position < valuesAsList.Count; position++)
                {
                    doOnValue?.Invoke(valuesAsList[position]);
                }
            }
        }
```
the foreach is a good way for code lisibility. And I have the advantage of for boost , with foreach lisibility.

**What I would like**
I'd like the foreach to behave like for loop, i.e not using IEnumerable, when dealing with IList.

Is it possible ?

Regards,

VGib

**Comment by YairHalberstadt** [comment]
This isn't always the case that for is faster than foreach.

For example looping over an ImmutableList is O(n) when using `foreach` but O(n * log(n)) when using `for`.

**Comment by VGib** [comment]
OK, I didn't do benchmark with ImmutableList, just my own use case.

Is it possible to do this when the for loop is known to be faster.

For my purpos, this is typically the case when the compilator should choose to change your code, without having to guess which is faster of for and foreach.

This may be to complicated for common developpers.

If this seems too complicated let's close it.

**Comment by YairHalberstadt** [comment]
How would the compiler know which is faster?

**Comment by VGib** [comment]
I've take a look to ImmutableList. 

This is obviously not a good idea, to use for loop. (But without knowing, I'll may use the for loop).

Maybe a #something before the class. This should be present on List and array. As they are lot used, this may be a good idea.

Any way: this need to have to be able to have information read at compile time. I don't know if this can be done.

If too complicated let's close this. But this seems to me great.

**Comment by YairHalberstadt** [comment]
But the type isn't known at compile time. At compile time all we have is an IEnumerable

**Comment by VGib** [comment]
OK, so false good idea.
Let's close it.

**Comment by benaadams** [comment]
In you example it would go faster if you also only evaulated `.Count` via the interface once
```csharp
int count = valuesAsList.Count;
for (int position = 0; position < count; position++)
{
    doOnValue?.Invoke(valuesAsList[position]);
}
```
Sprinkle in some pattern matching to combine the cast and null check
```csharp
public static void ForEach<T>(this IEnumerable<T> values, Action<T> doOnValue)
{
    if (values is IList<T> list)
    {
        int count = list.Count;
        for (int position = 0; position < count; position++)
        {
            doOnValue?.Invoke(list[position]);
        }
    }
    else
    {
        foreach (var value in values)
        {
            doOnValue?.Invoke(value);
        }
    }
}
```

**Comment by VGib** [comment]
yes, you're right. I'll fix it.
Thanks

**Comment by HaloFour** [comment]
In Java they used a marker interface `RandomAccess` to indicate that the collection was efficient to navigate by index.  `ArrayList<T>` implements it, but `LinkedList<T>` does not.  It's kind of a shame .NET didn't have similar.

https://docs.oracle.com/javase/8/docs/api/java/util/RandomAccess.html

**Comment by YairHalberstadt** [comment]
@HaloFour

But ImmutableList does have efficient random access. It's just slower than sequential access.

**Comment by VGib** [comment]
So if it's slower, it just don't have the interface (could be rename in IsFasterWithRandomAccess).

**Comment by HaloFour** [comment]
@YairHalberstadt 

Per the Java docs I linked the general rule of thumb is to apply that marker interface if it is expected that a `for` loop would be faster than using an iterator (or a `foreach` loop).  But it also explicitly calls out how fuzzy this can be.

**Comment by YairHalberstadt** [comment]
@HaloFour

But you also have to evaluate whether the cost of checking for that interface is greater than the benefits it gets you. Note that Java doesn't change the behaviour of a foreach loop to check for this interface.

**Comment by VGib** [comment]
I know that Linq is doing lot of check optimization to check if there is a IList or just a IEnumerable. 
For sample, the  [first method](https://github.com/microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs) : 
```csharp
    public static TSource First<TSource>(this IEnumerable<TSource> source) {
            if (source == null) throw Error.ArgumentNull("source");
            IList<TSource> list = source as IList<TSource>;
            if (list != null) {
                if (list.Count > 0) return list[0];
            }
            else {
                using (IEnumerator<TSource> e = source.GetEnumerator()) {
                    if (e.MoveNext()) return e.Current;
                }
            }
            throw Error.NoElements();
        }
```

The question is: is the average performance boost better ?

**Comment by YairHalberstadt** [comment]
@VGib

Linq doesn't do those checks when it's just planning to enumerate. It only does the checks if it can use them more effectively in some way. For example it can make first allocation free by checking for IList.

**Comment by HaloFour** [comment]
@YairHalberstadt 

> Note that Java doesn't change the behaviour of a foreach loop to check for this interface.

That they don't.  It's there strictly as a helper for people writing general APIs that need to iterate over `List<T>` as a `for` loop could be accidentally quadratic.

**Comment by YairHalberstadt** [comment]
Just to be clear, I am not saying this is necessarily a bad idea. What I'm saying is this needs way more evidence that it's a good idea.

For example, if you were to use a Roslyn based tool to rewrite CoreFX to check for the interface before looping, and then run some of their benchmarks, and there were significant speedups, that would be much more convincing that something is worth doing here.

**Comment by VGib** [comment]
@YairHalberstadt 

+1 a RandomAccess interface, with a roslyn warning would be nice.

Lot of people improve them self by using this kind of warning.

**Comment by VGib** [comment]
But I still don't use for loop, I prefer foreach for code lisibility. Even if I know, that for List for is faster.

I switch, only where their is performance purpose

**Comment by tannergooding** [comment]
There is no guarantee that `IEnumerable<T>` will enumerate items order from `0 to Count`, so I don't think this is a safe optimization for the compiler to make.

**Comment by VGib** [comment]
This is always the same debat.

their could be an interface to identify what prefer foreach, what could be done with for, without drawbacks.

I'm not as experienced as you. But I've been doing lot of assignment on lot of clients since 10 years.
For what I've seen. clients use a lot the List class, for some it's more than 90% of their foreach use.

For this case this seems to be a good idea to remplace foreach with for.
I'm aware, that there are case where this is not a good idea.

But maybe some good compromise could be found.

**Comment by HaloFour** [comment]
The compiler will already use an optimized version of an enumerator if a `GetEnumerator` method is present.  That can avoid allocation and virtual dispatch.  The team has [already considered](https://github.com/dotnet/csharplang/issues/1424) having `foreach` automatically behave like `for` if the target has `Count`/`Length` members and an indexer and deemed it unworthy of a language change.  I'd say that the likelihood of the BCL adding in a `RandomAccess`-like marker interface at this point is also quite unlikely to happen.  `for` and `foreach` are different, it's incumbent on the developer to know which to use and when.

At best I think an analyzer that identifies `foreach` over known list types and offering a code fix to convert to `for` would be the best bet.  You can suggest it at https://github.com/dotnet/roslyn-analyzers.

**Comment by tannergooding** [comment]
> But maybe some good compromise could be found.

Yes, it could. But it would require either:
* Some explicit syntax in the foreach, indicating the developer says it can be transformed
* Some explicit attribute on the interfaces/types indicating it is a valid transformation
  * This can't be readily added to the framework types, because we also don't know how the implementer will enumerate items either

At the end of the day, it's a breaking change and even if 90% of apps will "just work" with the change, that remaining 10% still needs to be considered, at which point (imo) just writing the few extra characters to use a `for` loop seems reasonable.

```cs
foreach (var item in list)
{
    // code
}

// vs

for (int i = 0; i < list.Count; i++)
{
    var item = list[i];
    // same code
}
```

**Comment by VGib** [comment]
what could be a nice idea is to have a syntax like this:
```csharp
for(var item in list)
{
  // do some stuff with item as with for loop
}
```

Any way this may be confusing, so I'm not sure it'll be a good idea.

The problem is that lot of people I know, prefer foreach to for loop for maintability purpose, and for loops are seen by many as premature optimization.
And when boosting their application, you can't turn all foreach into for.

**Comment by Thaina** [comment]
Isn't this #1424 ?

And I have said it there

> I don't think we need to generalize that
> 
> `Current` and `MoveNext` is seriously most generalized form of iteration. If there is object that should iterate by indexer then it's iterator job to optmize that by kept only index and compare with length for `MoveNext`
> 
> If returning iterator object is costly then I have seen a proposal on valuetype iterator #982
> If returning object from Current is costly then I have a proposal IRefEnumerable #2503 
> If checking for MoveNext is costly then I have proposal on undefinable #521 that should also work on foreach

**Comment by john-h-k** [comment]
This is a breaking change, and they are inherently different things

```cs
public class OhNo<T> : IList<T>, IEnumerable<T> where T : new() // implicit
{
    public IEnumerator<T> GetEnumerator() => null;
    public T this[int index] => new T();
    public int Count => int.MinValue;
}
```

Evaluating this using `Count` and the indexer is completely different behaviour to the enumerable. Sure, this is absolutely horrific code, but it is still a breaking change.

**Comment by VGib** [comment]
OK

**Comment by VGib** [comment]
Hello,

I was wondering something, excuse-me if I post my idea, when you have already answer.

Is it possible to make an optimisation on the Enumerable class, to prefer the for loop instead.

Because the List and Array have a special Enumerator ( from [https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/generic/list.cs](List repositoryl) )

```csharp
 [Serializable]
        public struct Enumerator : IEnumerator<T>, System.Collections.IEnumerator
        {
            private List<T> list;
            private int index;
            private int version;
            private T current;

            internal Enumerator(List<T> list) {
                this.list = list;
                index = 0;
                version = list._version;
                current = default(T);
            }

            public void Dispose() {
            }

            public bool MoveNext() {

                List<T> localList = list;

                if (version == localList._version && ((uint)index < (uint)localList._size)) 
                {                                                     
                    current = localList._items[index];                    
                    index++;
                    return true;
                }
                return MoveNextRare();
            }

            private bool MoveNextRare()
            {                
                if (version != list._version) {
                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_EnumFailedVersion);
                }

                index = list._size + 1;
                current = default(T);
                return false;                
            }

            public T Current {
                get {
                    return current;
                }
            }

            Object System.Collections.IEnumerator.Current {
                get {
                    if( index == 0 || index == list._size + 1) {
                         ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_EnumOpCantHappen);
                    }
                    return Current;
                }
            }
    
            void System.Collections.IEnumerator.Reset() {
                if (version != list._version) {
                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_EnumFailedVersion);
                }
                
                index = 0;
                current = default(T);
            }

        }
    }
}
````

This would be nice to have the roslyn warning on the Enumerator type, with maybe an interface IEnumerablePreferForLoop.

What do-you think ?

**Comment by HaloFour** [comment]
@VGib 

If that's what you want you can write an analyzer which warns on `foreach` over any specific implementation of `IEnumerable<T>` or `IList<T>`, and offer a fix to convert to a `for` loop instead. But that doesn't sound like something that the compiler should enforce.

**Comment by VGib** [comment]
Yes I'd like (maybe I would I like the CLR to optimize the loop with something like:
Get the enumerator, if Enumerator prefer the for loop, optimize with a for loop, but this may seems dreaming :) ).

But any way, this seems me much more easier to implement, than what we speek about. Since this is just an interface you put on Enumerator, and not on upper class.

**Comment by yaakov-h** [comment]
Using a for loop would lose version-safety, i.e. throwing an exception if the collection was modified whilst enumerating.

This is not something the compiler can just swap out for you.

**Comment by VGib** [comment]
OK so bad idea (once more).

**Comment by tannergooding** [comment]
It's not that it is a bad idea.

It's just that there are a lot of subtleties to any language proposal and unless you have previous experience with that, it is very easy to miss the various reasons why something might not work.

Everyone has pain points with every language and so everyone has ideas on how to fix those pain points. Putting up issues for those ideas is completely fine and the appropriate people (sometimes just other members from the community) will comment on why 'x' will or won't work, various edge cases that need to be considered, or link to existing/related proposals.

If people don't log an issue for the pain points they have, they language team may never know that some people consider it a pain point and so it or some feature to address it, may never come into existence (as it then requires someone from the language team to consider it a pain point, which they may not).

**Comment by avin-kavish** [comment]
@benaadams Isn't this a compiler optimizaton?

````c#
int count = list.Count;
for (int position = 0; position < count; position++)
{
}

**Comment by benaadams** [comment]
> @benaadams Isn't this a compiler optimizaton?

No, the compiler can't make that optimization.

What the interface call `IList.Count` does is unknown to the compiler; its also unknown to the Jit so it can't be hoisted out of the loop and must be called each time; as it could do anything,

You can make the choice to assume its well behaved and will return the same value each time so manually hoist it out of the loop; but neither compiler can make that assumption.

If its a direct call rather than a interface, or virtual call (like on an array, span or List); then the Jit will likely inline it an then hoist it out of the loop automatically; but it cannot do that with an interface dispatch as it cannot "see" what the method does.

**Comment by HaloFour** [comment]
@avin-kavish 

For a `LinkedList<T>` it becomes an accidentally quadratic "optimization", in that instead of walking the nodes once as it would with `foreach` using an `Enumerator<T>` it will walk all of the nodes up until the index for each iteration of the loop.

**Comment by CyrusNajmabadi** [comment]
> For a LinkedList<T> it becomes an accidentally quadratic "optimization", in that instead of walking the nodes once as it would with foreach using an Enumerator<T> it will walk all of the nodes up until the index for each iteration of the loop.

Note: a smart linked list coudl keep track of the last node/index that was looked up to make striding like this fast.  It's something i've seen in a lot of interesting datastructures (for example, string-apis with non-fixed-width-char lengths).

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

Sure, theoretically an implementation could, but the one in .NET doesn't and it's not wise to assume that any implementation would do that.  And ultimately something like that would just be more complicated than just using an enumerator since now the list has to track mutable state which would need to be synchronized across threads, etc.

**Comment by CyrusNajmabadi** [comment]
> Sure, theoretically an implementation could, but the one in .NET doesn't and it's not wise to assume that any implementation would do that. And ultimately something like that would just be more complicated than just using an enumerator since now the list has to track mutable state which would need to be synchronized across threads, etc.

Sure :)    Just throwing it out there for completeness :D

**Comment by avin-kavish** [comment]
@benaadams Thanks, yeah runtime optimization is what I meant. I was thinking of C++ compiler optimizations and forgot that C# needed a run-time for a second there.

> but it cannot do that with an interface dispatch as it cannot "see" what the method does.

Doesn't it "see" after the first dispatch? Once the procedure that it invoked does it's task it must be able to infer and optimize following calls to it?

@HaloFour  wow, that's gonna give me a nightmare. xD
